extensions/Worker.Extensions.Storage.Blobs/src/BlobStorageConverter.cs

```diff
@@ -135,72 +93,143 @@ private bool IsBlobExtension(ModelBindingData bindingData)
-            if (string.IsNullOrEmpty(connectionName))
+            if (string.IsNullOrEmpty(blobData.ContainerName))
             {
-                throw new ArgumentNullException(nameof(connectionName));
+                throw new ArgumentNullException(nameof(blobData.ContainerName));
             }
 
-            if (string.IsNullOrEmpty(containerName))
+            BlobContainerClient container = CreateBlobContainerClient(blobData.Connection!, blobData.ContainerName!);
+
+            if (IsCollectionBinding(targetType, blobData.BlobName!))
             {
-                throw new ArgumentNullException(nameof(containerName));
+                return await BindToCollectionAsync(targetType, container);
             }
+            else
+            {
+                if (targetType == typeof(BlobContainerClient) && !string.IsNullOrEmpty(blobData.BlobName))
+                {
+                    throw new InvalidOperationException("Binding to BlobContainerClient with a blob path is not supported. "
+                                                        + "Either bind to the container path, or use BlobClient instead.");
+                }
 
-            return await ToTargetTypeAsync(targetType, connectionName, containerName, blobName);
+                return await ToTargetTypeAsync(targetType, container, blobData.BlobName!);
+            }
         }
 
-        private async Task<object?> ToTargetTypeAsync(Type targetType, string connectionName, string containerName, string blobName) => targetType switch
-        {
-            Type _ when targetType == typeof(string) => await GetBlobStringAsync(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(Stream) => await GetBlobStreamAsync(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(byte[]) => await GetBlobBinaryDataAsync(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(BlobBaseClient) => CreateBlobClient<BlobBaseClient>(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(BlobClient) => CreateBlobClient<BlobClient>(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(BlockBlobClient) => CreateBlobClient<BlockBlobClient>(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(PageBlobClient) => CreateBlobClient<PageBlobClient>(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(AppendBlobClient) => CreateBlobClient<AppendBlobClient>(connectionName, containerName, blobName),
-            Type _ when targetType == typeof(BlobContainerClient) => CreateBlobContainerClient(connectionName, containerName),
-            _ => await DeserializeToTargetObjectAsync(targetType, connectionName, containerName, blobName)
+        private bool IsCollectionBinding(Type type, string blobName)
+        {
+            bool isContainer = string.IsNullOrEmpty(blobName);
```

Comment:
When the binding is `container/subdir`, blobName will be `subdir`.

We can either use the blobPath instead (will need this code change in first #1660). Or we can have our own rule where there has to be a trailing `/` i.e. `container/subdir/`

Reply:
Second thoughts on the training `/` we would have to make another WebJobs change as this will fail the blob path validation